File: forwardlist.cpp
#include <iostream>
#include <string>
#include <fstream>
#include "forwardlist.h"

using namespace std;

FNode* createNode(string data, FNode* next) {
    return new FNode{data, next};
}

void createFL(forwardList& fList, string data) { 
    FNode* firstNode = createNode(data, nullptr);
    fList.head = firstNode;
    fList.tail = firstNode;
}

void addAtHead(forwardList& fList, string data) {
    if (fList.head == nullptr) {
        createFL(fList, data);
    }
    else {
        FNode* newHead = createNode(data, fList.head);
        fList.head = newHead;
    }
}

void addAtTail(forwardList& fList, string data) {
    if (fList.head == nullptr) {
        createFL(fList, data);
    }
    else {
        FNode* newNode = createNode(data, nullptr);
        fList.tail -> next = newNode;
        fList.tail = newNode;
    }
}

void searchNode(const forwardList& fList, string targetValue) {
    if (fList.head == nullptr) {
        cout << "Список пуст.\n";
        return;
    }
    
    bool isFound = false;
    FNode* curElem = fList.head;
    int index = 0;

    while (curElem != nullptr) {
        if (curElem -> value == targetValue) {
            isFound = true;
            break;
        }
            
        curElem = curElem -> next;
        index++;
    }

    if (!isFound) {
        cout << "Элемент не найден.\n";
        return;
    }
    else {
        cout << "Элемент " << targetValue << " найден. Индекс: " << index << endl;
        return;
    }
}

void addBeforeIndex(forwardList& fList, string data, int index) {
    if (fList.head == nullptr) {
        cout << "Список пуст.";
        return;
    }
    else if (index == 0) {
        addAtHead(fList, data);
        return;
    }
    else if (index < 0) {
        cout << "Неверный индекс!\n";
        return;
    }

    int i = 0;
    FNode* curElem = fList.head;

    for (; i != index - 1 && curElem != nullptr; i++) {
        curElem  = curElem->next;
    }

    if (curElem == nullptr) {
        cout << "Элемента с таким индексом не существует!\n";
        return;
    }

    FNode* newNode = createNode(data, curElem->next);
    curElem->next = newNode;
}

void addAfterIndex(forwardList& fList, string data, int index) {
    if (fList.head == nullptr) {
        cout << "Список пуст.";
        return;
    }
    else if (index < 0) {
        cout << "Неверный индекс!\n";
        return;
    }

    FNode* curElem = fList.head;
    int i = 0;

    for (; i != index && curElem != nullptr; i++) {
        curElem = curElem->next;
    }

    if (curElem == nullptr) {
        cout << "Элемента с таким индексом не существует!\n";
        return;
    }

    FNode* newNode = createNode(data, curElem->next);
    curElem->next = newNode;

    if (curElem == fList.tail) {
        fList.tail = newNode;
    }
}

void deleteAtHead (forwardList& fList) {
    if (fList.head == nullptr) {
        cout << "Нечего удалять.\n";
        return;
    }
    
    FNode* deleteNode = fList.head;
    fList.head = fList.head -> next;

    if (fList.head == nullptr) {
        fList.tail = nullptr;
    }

    delete deleteNode;
}

void deleteAtTail(forwardList& fList) {
    if (fList.head == nullptr) {
        cout << "Список пуст.\n";
        return;
    }
    else if (fList.head == fList.tail) {
        deleteAtHead(fList);
        return;
    }
    
    FNode* curElem = fList.head;
    
    while (curElem -> next != fList.tail) {
        curElem = curElem -> next;
    }
    
    delete fList.tail;
    fList.tail = curElem;
    fList.tail -> next = nullptr;
}

void deleteBeforeIndex(forwardList& fList, int index) {
    if (fList.head == nullptr || fList.head == fList.tail) {
        cout << "Удаление невозможно.\n";
        return;
    }
    else if (index == 1) {
        deleteAtHead(fList);
        return;
    }
    else if (index <= 0) {
        cout << "Неверный индекс.\n";
        return;
    }

    FNode* curElem = fList.head;
    int i = 0;

    for (; i != index - 2 && curElem != nullptr; i++) {
        curElem = curElem->next;
    }

    if (curElem == nullptr || curElem == fList.tail || curElem->next == nullptr) {
        cout << "Неверный индекс.\n";
        return;
    }
    
    FNode* deleteNode = curElem->next;
    curElem->next = deleteNode->next; 
    delete deleteNode;
}

void deleteAfterIndex(forwardList& fList, int index) {
    if (fList.head == nullptr || fList.head == fList.tail) {
        cout << "Удаление невозможно.\n";
        return;
    }
    else if (index < 0) {
        cout << "Неверный индекс.\n";
        return;
    }
    
    FNode* curElem = fList.head;
    int i = 0;

    for (; i != index && curElem != nullptr; i++) {
        curElem = curElem->next;
    }

    if (curElem == nullptr || curElem == fList.tail) {
        cout << "Неверный индекс.\n";
        return;
    }

    if (curElem -> next == fList.tail) {
        deleteAtTail(fList);
        return;
    }

    FNode* deleteNode = curElem -> next;
    curElem -> next = deleteNode -> next;

    if (deleteNode == fList.tail) {
        fList.tail = curElem;
    }

    delete deleteNode;
}

void valueDelete(forwardList& fList, string data) {
    if (fList.head == nullptr) {
        cout << "Удаление невозможно.\n";
        return;
    } 

    FNode* curElem = fList.head;
    FNode* prevElem = nullptr;
    bool isFound = false;
    
    while (curElem != nullptr) {
        if (curElem->value == data) {
            isFound = true;
            break;
        }

        prevElem = curElem;
        curElem = curElem -> next;
    }

    if (!isFound) {
        cout << "Элемент не найден.\n";
        return;
    }

    if (curElem == fList.head) {
        deleteAtHead(fList);
    }
    else if (curElem == fList.tail) {
        deleteAtTail(fList);
    }
    else {
        prevElem->next = curElem->next;
        delete curElem;
    }
}

void printFL(FNode* fList) {
    cout << "Односвязный список: ";

    while (fList != nullptr) {
        cout << fList -> value << " ";
        fList = fList -> next;
    }

    cout << endl;
}

File: forwardlist.go
package main

import "fmt"

type FNode struct {
    value string
    next  *FNode
}

type ForwardList struct {
    head *FNode
    tail *FNode
}

func flCreateNode(data string, next *FNode) *FNode {
    return &FNode{value: data, next: next}
}

func createFL(data string) ForwardList {
    firstNode := flCreateNode(data, nil)
    return ForwardList{
        head: firstNode,
        tail: firstNode,
    }
}

func flAddAtHead(fList *ForwardList, data string) {
    if fList.head == nil {
        *fList = createFL(data)
    } else {
        newHead := flCreateNode(data, fList.head)
        fList.head = newHead
    }
}

func flAddAtTail(fList *ForwardList, data string) {
    if fList.head == nil {
        *fList = createFL(data)
    } else {
        newNode := flCreateNode(data, nil)
        fList.tail.next = newNode
        fList.tail = newNode
    }
}

func flSearchNode(fList ForwardList, targetValue string) {
    if fList.head == nil {
        fmt.Println("Список пуст.")
        return
    }

    isFound := false
    curElem := fList.head
    index := 0

    for curElem != nil {
        if curElem.value == targetValue {
            isFound = true
            break
        }
        curElem = curElem.next
        index++
    }

    if !isFound {
        fmt.Println("Элемент не найден.")
        return
    } else {
        fmt.Printf("Элемент %s найден. Индекс: %d\n", targetValue, index)
        return
    }
}

func flAddBeforeIndex(fList *ForwardList, data string, index int) {
    if fList.head == nil {
        fmt.Println("Список пуст.")
        return
    } else if index == 0 {
        flAddAtHead(fList, data)
        return
    } else if index < 0 {
        fmt.Println("Неверный индекс!")
        return
    }

    i := 0
    curElem := fList.head

    for i != index-1 && curElem != nil {
        curElem = curElem.next
        i++
    }

    if curElem == nil {
        fmt.Println("Элемента с таким индексом не существует!")
        return
    }

    newNode := flCreateNode(data, curElem.next)
    curElem.next = newNode
}

func flAddAfterIndex(fList *ForwardList, data string, index int) {
    if fList.head == nil {
        fmt.Println("Список пуст.")
        return
    } else if index < 0 {
        fmt.Println("Неверный индекс!")
        return
    }

    curElem := fList.head
    i := 0

    for i != index && curElem != nil {
        curElem = curElem.next
        i++
    }

    if curElem == nil {
        fmt.Println("Элемента с таким индексом не существует!")
        return
    }

    newNode := flCreateNode(data, curElem.next)
    curElem.next = newNode

    if curElem == fList.tail {
        fList.tail = newNode
    }
}

func flDeleteAtHead(fList *ForwardList) {
    if fList.head == nil {
        fmt.Println("Нечего удалять.")
        return
    }

    fList.head = fList.head.next

    if fList.head == nil {
        fList.tail = nil
    }
}

func flDeleteAtTail(fList *ForwardList) {
    if fList.head == nil {
        fmt.Println("Список пуст.")
        return
    } else if fList.head == fList.tail {
        flDeleteAtHead(fList)
        return
    }

    curElem := fList.head

    for curElem.next != fList.tail {
        curElem = curElem.next
    }

    fList.tail = curElem
    fList.tail.next = nil
}

func flDeleteBeforeIndex(fList *ForwardList, index int) {
    if fList.head == nil || fList.head == fList.tail {
        fmt.Println("Удаление невозможно.")
        return
    } else if index == 1 {
        flDeleteAtHead(fList)
        return
    } else if index <= 0 {
        fmt.Println("Неверный индекс.")
        return
    }

    curElem := fList.head
    i := 0

    for i != index-2 && curElem != nil {
        curElem = curElem.next
        i++
    }

    if curElem == nil || curElem == fList.tail || curElem.next == nil {
        fmt.Println("Неверный индекс.")
        return
    }

    deleteNode := curElem.next
    curElem.next = deleteNode.next
}

func flDeleteAfterIndex(fList *ForwardList, index int) {
    if fList.head == nil || fList.head == fList.tail {
        fmt.Println("Удаление невозможно.")
        return
    } else if index < 0 {
        fmt.Println("Неверный индекс.")
        return
    }

    curElem := fList.head
    i := 0

    for i != index && curElem != nil {
        curElem = curElem.next
        i++
    }

    if curElem == nil || curElem == fList.tail {
        fmt.Println("Неверный индекс.")
        return
    }

    if curElem.next == fList.tail {
        flDeleteAtTail(fList)
        return
    }

    deleteNode := curElem.next
    curElem.next = deleteNode.next

    if deleteNode == fList.tail {
        fList.tail = curElem
    }
}

func flValueDelete(fList *ForwardList, data string) {
    if fList.head == nil {
        fmt.Println("Удаление невозможно.")
        return
    }

    curElem := fList.head
    var prevElem *FNode = nil
    isFound := false

    for curElem != nil {
        if curElem.value == data {
            isFound = true
            break
        }
        prevElem = curElem
        curElem = curElem.next
    }

    if !isFound {
        fmt.Println("Элемент не найден.")
        return
    }

    if curElem == fList.head {
        flDeleteAtHead(fList)
    } else if curElem == fList.tail {
        flDeleteAtTail(fList)
    } else {
        prevElem.next = curElem.next
    }
}

func printFL(head *FNode) {
    fmt.Print("Односвязный список: ")

    for head != nil {
        fmt.Print(head.value + " ")
        head = head.next
    }

    fmt.Println()
}
